README.md
# XML Data Integrity Checksum (XDIC)

## Overview

The XML Data Integrity Checksum (XDIC) project provides a standard and reference implementation for embedding checksums in XML files. This approach is designed to ensure the integrity of XML data without altering its schema or content structure. XDIC uses cryptographic hash functions to generate checksums, which are then embedded within the XML files using Processing Instructions (PI).

## Purpose

The primary goal of XDIC is to provide a reliable method for verifying the integrity of XML data across different systems and platforms. This is particularly useful in environments where XML data is exchanged or stored over extended periods, and where maintaining its originality and integrity is crucial.

## Features

- Embeds checksums directly into XML files using PIs.
- Offers a standardized way to verify the integrity of XML data.
- Utilizes widely accepted cryptographic hash functions (e.g., SHA-256).
- Ensures that the XML data hasn't been tampered with or altered.
- Simple and straightforward implementation, similar to the [W3C's Subresource Integrity](https://www.w3.org/TR/SRI/)
- Does not require XML content normalization, ensuring simplicity and performance.

## Getting Started

To begin using XDIC in your projects, clone this repository to your local machine:

    git clone https://github.com/dave1010/XML-Data-Integrity-Checksum.git
    cd XML-Data-Integrity-Checksum

Check out the `src` directory for the reference implementation and the `examples` directory for sample XML files with embedded checksums.

## How It Works

XDIC embeds a [Processing Instruction](https://en.wikipedia.org/wiki/Processing_Instruction) in the XML file containing the checksum value. The checksum covers the entire XML file, excluding the checksum value itself. This way, any alteration in the XML content after the checksum generation invalidates the checksum, signaling potential tampering or corruption.

## Usage

XML files with XDIC include a processing instruction like this:

    <?xdic integrity="sha256-abc123..." normalize="false" version="1.0"?>

This is ignored by systems that do not understand the instruction. The checksum can be used to validate the rest of the XML file. The checksum can also be stored and referenced outside of the XML file.

Refer to the `examples` directory to see how the checksum PI is embedded in XML files. The `src` directory contains the reference implementation showing how to generate and verify checksums.

For detailed information on the checksum generation and verification process, please refer to the [SPECIFICATION.md](SPECIFICATION.md) file.

## Frequently Asked Questions

See the [Frequently Asked Questions](FAQs.md) page.

## Contributing

We welcome contributions to the XDIC project! Whether you're interested in fixing bugs, adding new features, or improving documentation, your help is appreciated. Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct and the process for submitting pull requests.

## License

This project is licensed to the public domain under CC0 1.0  - see the [LICENSE](LICENSE) file for details.

## Changelog

For a detailed list of changes and updates, please refer to the [CHANGELOG.md](CHANGELOG.md) file.

## Author

- [Dave Hulbert](https://dave.engineer) - Initial work and ongoing maintenance


## Acknowledgments

- Thanks to contributors and community members who have provided valuable feedback and suggestions.
- A tip of the hat to those brave souls who dive into XML files with nothing but a text editor and a dream. Your adventurous spirit has been a key inspiration for this project.




CHANGELOG.md
# Changelog

All notable changes to the XDIC (XML Data Integrity Checksum) project will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/), and this project adheres to [Semantic Versioning](https://semver.org/).

## [1.0] - 2023-11-16

### Added

- Initial release of the XML Data Integrity Checksum (XDIC) specification.
- Specification includes:
  - Definition and structure of the XDIC Processing Instruction (PI).
  - Detailed process for checksum generation and verification.
  - Guidelines for versioning and handling of different specification versions.
  - Considerations for compatibility and interoperability across various XML parsers and systems.
  - Security considerations discussing the scope and limitations of XDIC.
- FAQs document providing answers to common questions about XDIC.
- Reference implementation of the XDIC specification (link to implementation).
- `README.md`, `CONTRIBUTING.md`, and `LICENSE` files for project documentation and contribution guidelines.

### Notes

- This is the first official release of XDIC, marking the launch of the project.
- As this is the initial version, users and developers are encouraged to provide feedback and suggestions for future improvements.


xdic-web-tool/tests/testVerifyXdic.js
const { verifyXdicHash } = require('../js/verifyXdic.js');

function testVerifyXdicHash() {
    // Example XML content with a valid XDIC hash
    const validXml = `<?xml version="1.0" encoding="UTF-8"?>
<?xdic integrity="sha256-47689def1604cbb3ee7b70e3dff7d1c9115753083b41b987c063b1cc771b988a" version="1.0" normalize="false"?>
<rootElement>
    <!-- XML content goes here -->
</rootElement>`;

    // Example XML content with an invalid XDIC hash
    const invalidXml = `<?xml version="1.0" encoding="UTF-8"?>
<?xdic integrity="sha256-nope" version="1.0" normalize="false"?>
<rootElement>
    <!-- XML content goes here -->
</rootElement>`;

    // Test with valid XML
    verifyXdicHash(validXml)
        .then(result => console.log('Test with valid XML should pass:', result))
        .catch(error => console.error('Test with valid XML should not fail:', !error));

    // Test with invalid XML
    verifyXdicHash(invalidXml)
        .then(result => console.log('Test with invalid XML should not pass:', !result))
        .catch(error => console.error('Test with invalid XML should fail:', !error));
}

testVerifyXdicHash();


xdic-web-tool/tests/testGenerateXdicHash.js
const { generateXdicHash } = require('../js/generateXdic.js');
const { verifyXdicHash } = require('../js/verifyXdic.js');

function testGenerateXdicHash() {
    const testXml = `<?xml version="1.0" encoding="UTF-8" ?>
    <rootElement>
        <child>Sample Content</child>
        <!-- Comment -->
    </rootElement>`;

    // Test generating XDIC hash
    generateXdicHash(testXml)
        .then(generatedXml => {
            console.log('XDIC Hash generated successfully.');
            console.log('Generated XML:', generatedXml);

            // Optionally, verify the generated hash
            return verifyXdicHash(generatedXml);
        })
        .then(verificationResult => {
            if (verificationResult) {
                console.log('Verification of generated hash passed:', verificationResult);
            } else {
                console.error('Verification of generated hash failed:', verificationResult);
            }
        })
        .catch(error => console.error('XDIC Hash generation failed:', error));
}

testGenerateXdicHash();


xdic-web-tool/js/app.js
document.getElementById('upload-button').addEventListener('click', () => {
    document.getElementById('file-input').click();
});

document.getElementById('file-input').addEventListener('change', handleFileSelect);
document.getElementById('drop-zone').addEventListener('drop', handleFileDrop);

function handleFileSelect(event) {
    const file = event.target.files[0];
    if (file) {
        processFile(file);
    }
}

function handleFileDrop(event) {
    event.preventDefault();
    const file = event.dataTransfer.files[0];
    if (file) {
        processFile(file);
    }
}

function processFile(file) {
    const reader = new FileReader();
    reader.onload = function(e) {
        const content = e.target.result;
        verifyXdicHash(content)
            .then(isValid => {
                document.getElementById('verification-result').textContent = isValid ? 'Valid XDIC Hash.' : 'Invalid or no XDIC Hash.';
            })
            .catch(error => {
                document.getElementById('verification-result').textContent = 'Error: ' + error;
                document.getElementById('add-xdic-button').hidden = false;
            });

        document.getElementById('add-xdic-button').addEventListener('click', () => {
            generateXdicHash(content)
                .then(updatedXml => {
                    downloadFile(updatedXml, file.name);
                })
                .catch(error => {
                    console.error('Error generating XDIC:', error);
                });
        });
    };
    reader.readAsText(file);
}

function downloadFile(data, filename) {
    const blob = new Blob([data], {type: 'text/xml'});
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    window.URL.revokeObjectURL(url);
    a.remove();
}

// Prevent default drag behaviors
['dragenter', 'dragover', 'dragleave', 'drop'].forEach(eventName => {
    document.getElementById('drop-zone').addEventListener(eventName, preventDefaults, false);
});

function preventDefaults(e) {
    e.preventDefault();
    e.stopPropagation();
}

xdic-web-tool/js/generateXdic.js
function generateXdicHash(xmlString, hashAlgorithm = 'SHA-256') {
    return new Promise((resolve, reject) => {
        try {
            const normalizedHashAlgorithm = hashAlgorithm.toLowerCase().replace('-', '');

            // Step 1: Prepare the XML Document
            // Ensure that the XDIC PI is not yet in the document
            const piRegex = /<\?xdic integrity="[^"]*" version="[^"]*" normalize="[^"]*"\?>/i;
            if (piRegex.test(xmlString)) {
                throw new Error("XDIC Processing Instruction already exists in the document.");
            }

            // Step 2: Insert Placeholder Processing Instruction
            // Find the XML declaration
            const xmlDeclRegex = /<\?xml [^?]*\?>/;
            const hasXmlDecl = xmlDeclRegex.test(xmlString);
            const xmlDeclMatch = hasXmlDecl ? xmlString.match(xmlDeclRegex)[0] : null;

            // Placeholder for checksum
            const placeholderChecksum = `${normalizedHashAlgorithm}-`;
            const placeholderPi = `\n<?xdic integrity="${placeholderChecksum}" version="1.0" normalize="false"?>`;

            // Insert the XDIC PI after the XML declaration (if present) or at the beginning
            const xmlWithPlaceholder = hasXmlDecl 
                ? xmlString.replace(xmlDeclMatch, `${xmlDeclMatch}${placeholderPi}`) 
                : `${placeholderPi}${xmlString}`;
            
            // Step 3: Serialize the XML Content
            // Already in string format

            // Step 4: Calculate the Checksum
            const encoder = new TextEncoder();
            const data = encoder.encode(xmlWithPlaceholder);
            const cryptoModule = (typeof window === 'undefined') ? require('crypto').webcrypto : window.crypto;

            cryptoModule.subtle.digest(hashAlgorithm, data)
                .then(hashBuffer => {
                    // Step 5: Encode the Checksum
                    const hashArray = Array.from(new Uint8Array(hashBuffer));
                    const checksum = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');

                    // Step 6: Finalize the XML Document
                    // Update the XDIC PI with the actual checksum
                    const finalXml = xmlWithPlaceholder.replace(`${normalizedHashAlgorithm}-`, `${normalizedHashAlgorithm}-${checksum}`);
                    resolve(finalXml);
                })
                .catch(error => reject(error.message));
        } catch (error) {
            reject(error.message);
        }
    });
}

// Conditional export
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { generateXdicHash };
} else {
    window.generateXdicHash = generateXdicHash;
}


xdic-web-tool/js/verifyXdic.js
/**
 * Verifies the XDIC hash of a given XML string.
 * @param {string} xmlString - The XML content as a string.
 * @returns {Promise<boolean>} - A promise that resolves to `true` if the hash is verified, `false` otherwise.
 */
function verifyXdicHash(xmlString) {
    return new Promise((resolve, reject) => {
        try {
            // Step 1: Locate and Extract the XDIC PI
            const piRegex = /<\?xdic integrity="([^"]+)" version="([^"]+)" normalize="([^"]+)"\?>/i;
            const matches = xmlString.match(piRegex);
            if (!matches) {
                throw new Error("XDIC Processing Instruction not found.");
            }

            const [fullPi, integrity, version, normalize] = matches;
            const [algorithm, checksumValue] = integrity.split('-');

            const algorithmMap = {
                'sha256': 'SHA-256',
                'sha384': 'SHA-384',
                'sha512': 'SHA-512'
            };

            // Verify version (allowing minor version updates within major version 1)
            const majorVersion = version.split('.')[0];

            if (majorVersion !== "1") {
                throw new Error("Unsupported XDIC major version. Only major version 1 is supported.");
            }

            const normalizedAlgorithm = algorithmMap[algorithm.toLowerCase()];
            if (!normalizedAlgorithm) {
                throw new Error("Unsupported hash algorithm.");
            }

            if (normalize !== "false") {
                throw new Error("Normalize attribute must be 'false'.");
            }

            // Step 2: Remove the Checksum from the XDIC PI
            const modifiedXmlString = xmlString.replace(fullPi, `<?xdic integrity="${algorithm}-" version="${version}" normalize="${normalize}"?>`);

            // Step 3: Serialize the XML Content
            // (already in string format, so no additional action required)

            // Step 4: Recompute the Checksum
            const encoder = new TextEncoder();
            const data = encoder.encode(modifiedXmlString);

            const compareChecksums = hashBuffer => {
                // Step 5: Compare Checksums
                const hashArray = Array.from(new Uint8Array(hashBuffer));
                const recomputedChecksum = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
                const isVerified = checksumValue === recomputedChecksum;

                resolve(isVerified);
            };

            crypto.subtle.digest(normalizedAlgorithm, data)
                    .then(compareChecksums)
                    .catch(error => reject(error.message));
        } catch (error) {
            reject(error.message);
        }
    });
}

if (typeof module !== 'undefined' && typeof module.exports !== 'undefined') {
    module.exports = { verifyXdicHash };
} else {
    window.verifyXdicHash = verifyXdicHash;
}
